File: control/haskell/src/Control/Pricing.hs
-- Control.Pricing — the pricing policy as truth (F5 witness).
-- Official source: https://api-docs.deepseek.com/quick_start/pricing
-- Peak Mon-Fri 01:00-04:00 and 06:00-10:00 UTC; everything else off-peak.
module Control.Pricing
  ( Decision (..)
  , isPeak
  , route
  ) where

import Data.Time.Calendar (DayOfWeek (..), dayOfWeek)
import Data.Time.Clock (UTCTime (..))

peakWindows :: [(Double, Double)]
peakWindows = [(1, 4), (6, 10)]

-- | True inside a weekday peak window (UTC).
isPeak :: UTCTime -> Bool
isPeak t = weekday && any inWindow peakWindows
  where
    weekday = dayOfWeek (utctDay t) `notElem` [Saturday, Sunday]
    h = realToFrac (utctDayTime t) / 3600
    inWindow (lo, hi) = h >= lo && h < hi

data Decision = Decision
  { target :: String
  , model  :: String
  , egress :: Bool
  , reason :: String
  } deriving (Eq, Show)

localKinds, flashKinds, proKinds :: [String]
localKinds = ["atomic", "audit", "ops", "ingest", "train"]
flashKinds = ["plan", "label", "census-draft", "tips", "vision", "coach"]
proKinds   = ["census-deep", "verify", "teacher", "plan-pro", "label-pro"]

-- | Deterministic provider sizing. A model never decides which model to use.
route :: String -> UTCTime -> Decision
route kind now
  | kind `elem` localKinds = Decision "local" "local" False "local-first"
  | kind `elem` flashKinds = Decision "deepseek-flash" "deepseek-flash" True (kind ++ " after egress; flash")
  | kind `elem` proKinds
  = if isPeak now
      then Decision "deepseek-flash" "deepseek-flash" True "pro forbidden at peak"
      else Decision "deepseek-v4-pro" "deepseek-v4-pro" True "pro off-peak"
  | otherwise = Decision "local" "local" False "default local"

File: control/internal/pricing/pricing.go
// Package pricing implements the platform pricing policy.
// Official source: https://api-docs.deepseek.com/quick_start/pricing
// Peak Mon-Fri 01:00-04:00 and 06:00-10:00 UTC; everything else off-peak (50% price).
package pricing

import "time"

var peakWindows = [][2]float64{{1, 4}, {6, 10}}

// IsPeak reports whether the instant is inside a weekday peak window (UTC).
func IsPeak(now time.Time) bool {
	if now.Weekday() == time.Saturday || now.Weekday() == time.Sunday {
		return false
	}
	h := float64(now.Hour()) + float64(now.Minute())/60 + float64(now.Second())/3600
	for _, w := range peakWindows {
		if h >= w[0] && h < w[1] {
			return true
		}
	}
	return false
}

// Decision is the deterministic provider verdict. A model never decides which model to use.
type Decision struct {
	Target string
	Model  string
	Egress bool
	Peak   bool
	Reason string
}

var localKinds = map[string]bool{"atomic": true, "audit": true, "ops": true, "ingest": true, "train": true}
var flashKinds = map[string]bool{"plan": true, "label": true, "census-draft": true, "tips": true, "vision": true, "coach": true}
var proKinds = map[string]bool{"census-deep": true, "verify": true, "teacher": true, "plan-pro": true, "label-pro": true}

// Route sizes a kind of work to a provider.
func Route(kind string, now time.Time) Decision {
	peak := IsPeak(now)
	switch {
	case localKinds[kind]:
		return Decision{Target: "local", Model: "qwen2.5-coder:64k", Peak: peak, Reason: "local-first"}
	case flashKinds[kind]:
		return Decision{Target: "deepseek-flash", Model: "deepseek-flash", Egress: true, Peak: peak, Reason: kind + " after egress; flash"}
	case proKinds[kind]:
		if peak {
			return Decision{Target: "deepseek-flash", Model: "deepseek-flash", Egress: true, Peak: true, Reason: "pro forbidden at peak"}
		}
		return Decision{Target: "deepseek-v4-pro", Model: "deepseek-v4-pro", Egress: true, Reason: "pro off-peak"}
	default:
		return Decision{Target: "local", Model: "qwen2.5-coder:64k", Peak: peak, Reason: "default local"}
	}
}
